### PR TITLE
Add threads reaper

### PR DESCRIPTION
**Pull request by amenzhinsky**
In some situations using threaded way of reporting errors when main process exits payloads are not delivered because all threads are killed straightaway.

For example rake tasks:
```ruby
Rollbar.use_threaded

task :foo do
  begin
    raise 'bar'
  rescue => error
    Rollbar.error(error)
  end
end
```

Another way is to add a method that would do something like this code, but it looks ugly for me: 
```ruby
begin
  tmp = Rollbar.configuration.use_async
  Rollbar.configuration.use_async = false
  yield
ensure
  Rollbar.configuration.use_async = tmp
end
```

### CODE REVIEW

**Comment by amenzhinsky** [comment]
I'd like to discuss it and when it's ok I'll add some spec and do some refactoring if it's needed

**Comment by jondeandres** [comment]
Hey @amenzhinsky, thank you very much for this PR. There is one point I'm concerned about: spawn the reaper always this class is loaded, even it's not used as async handler.

I think we should create this thread only if the `Rollbar::Delay::Thread` is being used. For this we can spawn the reaper thread if it's not spawned on first `.call()`. What do you think?

About refactoring, just take a look at https://codeclimate.com/github/rollbar/rollbar-gem/pull/405, since the spawn method is too long. We can split it into 3 methods: `spawn_threads_reaper`, `build_reaper_thread`, `configure_exit_handler` or similar.

We'll need to test this manually and we'll need specs, but this is a great work! Thanks

**Comment by amenzhinsky** [comment]
Hi @jondeandres, I updated the PG, I followed all your advises except  `build_reaper_thread`, I added `::reaper` method that memoizes the reaper thread and call it in `spawn_threads_reaper` because I need access to it in `::configure_exit_handler` it's implicit a little. What do you think?

File: lib/rollbar/delay/thread.rb [hunk 1]
```diff
@@ -1,9 +1,61 @@
...
+                  break
+                else
+                  thread.join
+                end
+              end
+            end
+        end
+
+        def spawn_threads_reaper
+          unless @spawned
```
Comment (jondeandres) [inline lib/rollbar/delay/thread.rb]: According to the Ruby style, it's prefered not wrap a method inside a `unless` block. Here you can do `return if @spawned`

File: lib/rollbar/delay/thread.rb [hunk 2]
```diff
@@ -1,9 +1,61 @@
...
+            @spawned = true
+
+            reaper # we spawn the reaper thread before configuring it
```
Comment (jondeandres) [inline lib/rollbar/delay/thread.rb]: I'd prefer if here we have `@reaper||= build_reaper_thread`. And you use `attr_reader :reaper` in the class definition.

**Comment by amenzhinsky** [comment]
@jondeandres done, please have a look

File: lib/rollbar/configuration.rb [hunk 3]
```diff
@@ -44,6 +44,7 @@ class Configuration
     attr_accessor :uncaught_exception_level
     attr_accessor :scrub_headers
     attr_accessor :sidekiq_threshold
+    attr_accessor :thread_timeout
```
Comment (jondeandres) [inline lib/rollbar/configuration.rb]: This name is confusing. Also, I'm not sure this should be something configurable. Maybe we can just use a constant for this.

**Comment by jondeandres** [comment]
@amenzhinsky the PR is great now. I will need to play a little with this and check everything is working as expected.

Feel free to write and fix the specs for this.

Thanks!

**Pull request closed without merging**